services/src/util/server.rs

```diff
@@ -294,3 +310,80 @@ pub(crate) fn log_server_info() -> Result<()> {
 pub async fn not_implemented_handler() -> HttpResponse {
     HttpResponse::NotImplemented().finish()
 }
+
+#[cfg(target_os = "linux")]
+pub struct SocketFd(pub RawFd);
+
+/// attach the connection's socket file descriptor to the connection data
+#[cfg(target_os = "linux")]
+pub fn connection_init(connection: &dyn Any, data: &mut Extensions) {
+    if let Some(sock) = connection.downcast_ref::<TcpStream>() {
+        let fd = sock.as_raw_fd();
+        if let Ok(fd) = NonZeroI32::try_from(fd) {
+            let fd = RawFd::from(fd);
+            data.insert(SocketFd(fd));
+        }
+    }
+}
+
+/// start a new task that monitors the request's socket file descriptor and tries to detect when the connection is closed.
+/// The returned join handle can be awaited to get notified when the connection is closed (or, if given, when the timeout is reached).
+// TODO: the socket file descriptor might be re-used by a subsequent connection before we notice that
+//       the connections was closed. We maybe need a global list of all open connections and their socket file descriptors
+//       or some other way to invalidate the socket file descriptor in the old connection when a new one arrives.
+//       idea: have a global map of sockets being monitored for connection close. When a new connection arrives: get its fd and
+//       check whether this fd is currently being monitored. If so: notify the monitor channel of the new connection via a channel.
+#[cfg(target_os = "linux")]
+pub fn connection_closed(req: &HttpRequest, timeout: Option<Duration>) -> BoxFuture<()> {
+    use futures::TryFutureExt;
+
+    const CONNECTION_MONITOR_INTERVAL_SECONDS: u64 = 1;
+
+    if let Some(fd) = req.conn_data::<SocketFd>() {
+        let fd = fd.0;
+        let handle = crate::util::spawn(async move {
+            let mut data = vec![];
+            let start = Instant::now();
+
+            while timeout.map_or(true, |t| start.elapsed() >= t) {
+                let r = nix::sys::socket::recv(fd, data.as_mut_slice(), MsgFlags::MSG_PEEK);
+
+                match r {
+                    Ok(0) | Err(Errno::EBADF) => {
+                        // the connection seems to be closed
+                        return;
+                    }
+                    _ => (), // the connection seems to be still valid
+                }
+
+                tokio::time::sleep(std::time::Duration::from_secs(
+                    CONNECTION_MONITOR_INTERVAL_SECONDS,
+                ))
+                .await;
+            }
+        });
+
+        // TODO: return `pending` on JoinError?
+        let handle = handle.unwrap_or_else(|_| ());
+
+        Box::pin(handle)
+    } else if let Some(timeout) = timeout {
+        Box::pin(tokio::time::sleep(timeout))
+    } else {
```

Comment:
if we have a timeout specified, we only wait for the timeout?

Reply:
no, only if we don't finde a file descriptor for the current request we fall back to the timeout

Reply:
Is this reasonable to fall back to a timeout?

Is is not the case that:
 * if we set a timeout, we want to use that timeout
 * also, we want to cancel on drop-out
 
?

Reply:
> Is this reasonable to fall back to a timeout?

well we discussed that we want to be able to specify a maximum time for requests. This method does both, abort on connection close or when the timeout is reached

> if we set a timeout, we want to use that timeout

the ogc protocols have no side-effects. It makes no sense to keep them running to a timeout if the connection is dropped anyway.

> 
    if we set a timeout, we want to use that timeout
    also, we want to cancel on drop-out


that's what I coded I think.

Reply:
For me, it looks like if you get a file descriptor, you start a future that waits for that descriptor being closed.

If you don't get a file descriptor, you start a future that waits for the timeout if this is available.

If you have neither an fd, nor a timeout, you just wait forever in a third future

Reply:
no, in line 348 you see that I also use the timeout when the file descriptor is being monitored. If no fd is present, fall back to the timeout. If no timeout is present, produce a future that never completes, thus the request will run until the query is finished. I think this is what we want.